src/background/util.ts

```diff
@@ -66,46 +42,126 @@
   throw new Error(`contentScript not ready in ${maxWaitMillis}ms`);
 }
 
+/** Checks whether a URL has permanent permissions and therefore whether `webext-dynamic-content-scripts` already registered the scripts */
+export async function isContentScriptRegistered(url: string): Promise<boolean> {
+  const { origins } = await getAdditionalPermissions({
+    strictOrigins: false,
+  });
+
+  return patternToRegex(...origins).test(url);
+}
+
+interface TargetState {
+  url: string;
+  installed: boolean;
+  ready: boolean;
+}
+
+/** Fetches the URL from a tab/frame. Return value is empty if it's unknowable (likely because we don't have permission to access it) */
+export async function getTargetState(
+  target: Target
+): Promise<TargetState | undefined> {
+  if (!isBackgroundPage()) {
+    throw new Error(
+      "getTargetState can only be called from the background page"
+    );
+  }
+
+  return browser.tabs
+    .executeScript(target.tabId, {
+      // This imitates the new chrome.scripting API by wrapping a function in a IIFE
+      code: `(${() => ({
+        url: location.href,
+        installed: Symbol.for("pixiebrix-content-script") in window,
+        ready: (window as any).pbReady,
```

Comment:
Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/pixiebrix/pixiebrix-extension/pullRequest?prid=7684115)
```suggestion
        ready: (window as unknown).pbReady,
```